### ISSUE TITLE
Workflow with develop and master

### ISSUE THREAD

**Issue by rljacob** [NCAR/ParallelIO#282]
The develop branch is in place but isn't being used correctly.  Probably because we didn't talk about how to use it.

develop is a "throw away" branch and can be wiped and remade at anytime.  As such, no feature branches should start from develop.
All feature branches should start from master and have a goal of making it back to master.
Because the goal of a feature branch is to make it to master, the PR should have master as the base.
The merge to develop for testing has to be done at the command line. 

Summary when starting from an updated master:
git checkout mybranch
(do development)
git push origin mybranch
Issue PR to master for mybranch

(in the same or a different clone)
get fetch origin (Make sure you have the latest changes!)
get checkout develop
git merge mybranch
(this would be  good point to run testing by hand on your merged develop branch)
git push origin develop
(wait for testing to run on develop;   If successful, you can use the green button to merge to master or use the above procedure substituting "master" for "develop")

If testing is NOT successful:
If you know the fix, commit it to your branch and re-merge the branch to develop.
If you can't fix it that day, revert the merge from develop

Also if testing is NOT successful, no other PRs can go in until the dashboard is green again.

**Comment by dqwu** [comment]
So I think we should make master the default branch again? When I check out PIO and use git status it shows the develop branch. I think users might not be aware of this (and they will have to use git checkout master)

**Comment by rljacob** [comment]
@jedwards4b please change the default branch back to master.

**Comment by dqwu** [comment]
Some new nightly builders should be added on develop branch (or triggered by merge commits to develop), right?

After a PR is approved, usually who should merge it to develop for testing? The owner of that PR (creator of the feature branch) or the integrator (who merges "graduated" PRs to master)?

**Comment by rljacob** [comment]
Since github now lets you mark a PR as "approved", the mechanics of actually merging it can be done by the PR owner.

**Comment by dqwu** [comment]
For merge a PR to develop it will be done via command lines. For merge a PR to master, should we use the green button or command lines? It seems that ACME workflow does not like the green button.

**Comment by rljacob** [comment]
We didn't like the green button because you couldn't edit the merge commit message.  Now you can so its ok.  (might need to change that in ACME).

**Comment by dqwu** [comment]
It is noticed that some PRs (like #278) are already merged to the develop branch and closed. Maybe that was done by the green button (develop branch becomes the default branch like master, and master branch downgrades to a regular branch, so the green button is only applied to develop)?

Some questions:
1) Should the develop branch be recreated (deleted and created again)?
2) What about the PRs that are merged to develop branch (possibly via the green button) and get closed? Should they be reopened such that they can be merged to master (will be the default branch again later) via the green button?

**Comment by jayeshkrishna** [comment]
Also note that the "green button" in github does not allow you to test your code (simple tests like "code compiles") after a merge (before committing it to repo).

**Comment by jedwards4b** [comment]
I feel like there is a lot of second guessing and speculation going on here by people who are minimally involved in development.

**Comment by rljacob** [comment]
We do tons of development with the develop/master workflow.  Just not here.  You're doing it wrong.

**Comment by dqwu** [comment]
PETSc (https://bitbucket.org/petsc/petsc) is using the workflow described by Rob, where the development branch for testing PRs is called "next". They have been using this for a long time.

MOAB (https://bitbucket.org/fathomteam/moab) uses the same workflow. It did not have the develop branch at the beginning, though.

**Comment by rljacob** [comment]
And both are based off of gitworkflows which is used by the Linux kernel.  ftp://www.kernel.org/pub/software/scm/git/docs/gitworkflows.html

**Comment by rljacob** [comment]
It occurred to be that when someone said "maybe we should use a develop or next branch" they weren't using "develop" as a synonym for "next".  There's a different workflow called "gitflows" which typically calls their two branches "master" and "develop".    However it has its own problems summarized here:  http://www.mail-archive.com/fenics@lists.launchpad.net/msg01396.html

**Comment by jedwards4b** [comment]
After the merge of branch yellowstone_pgi_update was successfully tested in develop I opened a PR to merge it to master and completed that merge.   I don't see what the problem is with having develop as the default branch.

**Comment by rljacob** [comment]
Well if you do a fresh clone of ParallelIO, you start on the "develop" branch instead of "master".  That will confuse people starting new work (if they don't check for it).   2 PRs per branch will clutter the open and closed PR list making it hard to tell the status or history of work.  You'll be constantly asking  "Is this open PR for the develop merge or the master merge?" and "has that feature had its develop PR already?" when there's a bunch of open PRs.

**Comment by edhartnett** [comment]
@jedwards4b I certainly take your point that there are a lot of people minimally involved in development who are putting forward a lot of rules! ;-)

I think it is worth discussing and trying @rljacob's suggested workflow. But Robert you must be patient while we get our heads around the new procedure.

Right now we have a broken build, and I am trying to get a fix together. I am changing my own test environment, and it's not working well at the moment. Once I am sure I have a good working environment ready, I will run tests on my changes, and submit them as a PR.

I think we should proceed to merge to develop and then to master, just as Jim has done, for the next couple of PRs, until I get the build working again.

Then let's try it Robert's way and see how it works in practice.

**Comment by rljacob** [comment]
Sounds good.

**Comment by jedwards4b** [comment]
There is no one right way to do things and no amount of process and procedure is going to overcome an antagonistic relationship between developers.  

I will change the cdash procedure to checkout the develop branch explicitly, I can then change the default back to master.

**Comment by edhartnett** [comment]
Well we have restarted the process and are trying to proceed as outlined by Robert.

It has been a busy day, I put up 3 PRs, Jim 2, and Jayesh 1.

Review seems to be working well, and we are all putting up small and isolated changes that are easy to review. We are not leaving them up "about a day" but are instead merging them as soon as approved.

I think that's working fine, and we should change our rules to reflect that. But we also have to be aware that sometimes that won't work. One reason it is working today is that we are all staying in our patches. 

But there are no changes to the C library that I don't want to review before merging. Be advised.

We are each keeping track of when our changes to develop pass cdash, and then merging our commits to master. I believe this is ripe for human error. I can easily see forgetting the last step, and fixing something in develop, and forget to merge to master. @rljacob how do you avoid that on the other projects using this process?

Have you guys changed your Jenkins to build develop? And also master? Do you build with a variety of options?

I think merging first to develop, and then to master, is not yielding any obvious benefit.

**Comment by dqwu** [comment]
On our Jenkins server (https://jenkins-ci.mcs.anl.gov):
ParallelIO_valgrind_develop builds develop and generate valgrind check report
ParallelIO_valgrind builds master and generate valgrind check report
ParallelIO_coverage builds master and generate code coverage report

We will add more builders later.

**Comment by jedwards4b** [comment]
But these don't run any tests?

**Comment by dqwu** [comment]
_"But these don't run any tests?"_
They all run tests by make check. You can view the console output for more information.

_"I think merging first to develop, and then to master, is not yielding any obvious benefit."_
We need sufficient nightly builders on CDash/Jenkins to build develop branch. If new build errors or test failures are observed after the feature branch of a PR is merged to develop, the master branch will be protected as that PR has not been merged to master yet. This is helpful to maintain a stable master branch.

**Comment by edhartnett** [comment]
Actually I was interupted there, and didn't get to finish my thought, which
was that I didn't see any obvious benefit, but that I expect over time, the
extra stability of the master branch will be apparent. Anyway, I am
certainly willing to continue with the new process.

I am a little worried about human error, with an extra step to get changes
to master. I will try and work out some good habits there.

**Comment by edhartnett** [comment]
Here's a problem I am having this morning with the new workflow.

This morning, I started another task, on a new branch, which I took from master, not develop, as instructed by @rljacob. The new branch is ejh_check_netcdf2.

I did a bunch of changes on ejh_check_netcdf2 and all tests are passing. I would like to put up a PR for ejh_check_netcdf2 changes.

But when I do this on github, I put up a PR against develop, and that shows not just my ejh_check_netcdf2 changes, but also attempting to undo all the changes against develop that have not yet merged to master. (Because I took my branch from master, not develop, and things have been merged to develop that are not yet on master.)

How do I get a PR against develop showing only the changes in my branch, and ignoring the differences between master and develop that have nothing to do with my changes?

**Comment by edhartnett** [comment]
@rljacob when I look around on the web, I do not see anyone describing a master/develop workflow with git that you guys have described to us. In particular, I don't see anyone else advising taking feature branches off master; also everyone seems to accept PRs from develop to master in order to bring changes to the master branch. Can you show me any on-line descriptions of this workflow?

I have worked with gitflow https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow. In that case, we take our feature branches from develop.

I wonder if we would not be better off going back to feature branch workflow (https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) with an automated, tested, daily snapshot release. That ensures a stable build every day, without imposing so much extra work on me.

If a more complex structure is required, can we go with gitflow? It seems more straightforward and better-documented than the process you describe.

For the last two or three days I have been working at about 25% my normal rate, due to all the issues with the new process. If it does not smooth out soon, we will have to make some changes. At the moment I am stuck because there is no way to show my changes in a PR.

**Comment by rljacob** [comment]
The PR should be against "master".  Don't make a PR for the merge to develop, do that as a command line step.

In this workflow, "develop" is usually called "next" so search for "master/next".  Its also somewhat confusingly called "gitworkflows".    There are  variations but its used by several groups including the Linux kernel.
https://git-scm.com/docs/gitworkflows
ftp://www.kernel.org/pub/software/scm/git/docs/gitworkflows.html
https://bitbucket.org/petsc/petsc/wiki/developer-instructions-git

The many problems with "gitflow" are summarized here:
http://www.mail-archive.com/fenics@lists.launchpad.net/msg01396.html

**Comment by edhartnett** [comment]
If I don't do a PR for merge to develop, how does my code get reviewed?

**Comment by jedwards4b** [comment]
I suppose that it gets reviewed after it's on the develop branch and running tests.

**Comment by edhartnett** [comment]
I am not happy with that plan at all. 

@rljacob do you guys do code review on the other projects that use this process? When does the code get reviewed?

**Comment by jayeshkrishna** [comment]
The PR is created against master. The code reviewer then reviews the code. Once approved, the branch is merged to next. We wait for the testing to complete, and if everything looks good the branch is merged to master.
A tag for the PR is added to commit message when merging to next to link the commit with the PR

**Comment by edhartnett** [comment]
What happens if the code can merge cleanly into master, but conflicts with other changes in develop?

**Comment by jayeshkrishna** [comment]
In that case you would have to resolve the conflicts manually (the merge to next/develop is always a manual process) and merge to next/develop.

next/develop is not considered as important as master (the workflow's goal is to keep master clean). Ideally you wouldn't have master/next to be that different from each other.

**Comment by edhartnett** [comment]
For example let's say I take a branch and make some changes.

Meanwhile, Jim takes a branch and makes changes which overlap. He finishes first, takes a PR against master, gets reviewed and approved, and merges to develop for testing. (But still not master).

I PR my code against master and we review that and approve it. I try to merge to develop, but there are conflicts. So I have to make changes in the code (some Jim's, some mine), to get the merge to work.

This means that I am modifying my code after review? And those changes are not reviewed?

I then test the modified code (in develop), and it works. So I then merge my PR to master. So I am merging untested code to master? Because I had to modify the code to get it to work in develop, but that's not in my PR against master.

Then Jim comes along and tried to merge his PR to master, but it no longer works, because of my changes. So now he does what? Re-does the merge changes I already figured out (but are only on the "throw-away" develop branch?)

I must be missing something, because that does not sound right. Tell me what I got wrong in the example above...

**Comment by jayeshkrishna** [comment]
I think this is a general issue with two developers working on the same code. In the case of a major conflict (like the one that you mentioned) the second developer merging to next would have to wait till the first developer's changes gets to master, create a new branch that merges the second developer's branch and the master (after the first developer's changes goes in) and get it reviewed again (and merge this new branch).
The developers would have to coordinate and get the issue resolved.

**Comment by rljacob** [comment]
In any workflow, if you have to resolve conflicts before merging to develop or master, how are those reviewed?  That's not unique to gitworkflows.

**Comment by rljacob** [comment]
In your example, Jim started the PR first so he gets to go to next first and also to master first.  You will resolve conflicts the same way as you merge to next and to master.

Jed Brown, a git expert, is a professor at CU and helped define the ACME and PETSc git procedures.   I think he'd be happy to meet with you guys to go over all the scenarios and pros and cons.

**Comment by edhartnett** [comment]
@rjacob resolving the conflicts "the same way" when I merge to master is not good enough. Conflicts can be significantly complex, and if I can't merge directly from develop, I must re-do changes "by hand." So where's the guarantee I will do them the same, and not make mistakes?

In our previous workflow, conflicts were resolved without waiting, and will full review, once.

I'm sure this is a wonderful process for some other projects that don't have the same review needs we do. And it seems to be suited for projects with large numbers of contributors. But I suggest that we do not use it in this project. 

Instead I suggest we go back to the feature-branch development model we were using before. If master is thought to be too unstable, let us mandate Jenkins-based PR testing. No PR can be merged until it passes tests. The NetCDF project uses this process and it is very successful and convenient, both for the internal team, and in dealing with external contributors.

Jim, what do you think? I am finding this new process to be a significant drag on productivity, and I'm not sure it fits well with our existing review culture. Let's switch back to what we had before, and investigate other ways to improve the stability of the master branch.

**Comment by rljacob** [comment]
"In our previous workflow, conflicts were resolved without waiting, and will full review, once."  How did conflict resolution get reviewed in the previous workflow?

**Comment by edhartnett** [comment]
I think the workflow you are using is better suited to larger code bases, with more people, where conflicts are less likely.

@rljacob in the feature branch workflow, conflicts are dealt with by the second programmer when merging, and before code review. So conflict resolution is done once, without waiting, and code is not submitted to review until conflicts are resolved.

Another master/develop workflow I have used (which is what I thought I was signing up for), is very much simpler.

Developers work on the develop branch, just as we did on the master branch in the past. Periodically, after it passes all test, a merge is done from develop to master, with all changes. (Usually the authority to merge to master is limited so programmers don't do it by mistake.)

Code is reviewed as it goes into develop. It is never changed when merged to master, so is not re-reviewed. Only code that is fully-tested and fully-reviewed is merged to master.

That way, master is always stable, development continues on develop. No waiting, no unreviewed code.

**Comment by edhartnett** [comment]
BTW this is the first project I have ever used the GitHub API for this sort of thing, and I really like it. 

I think it is much less likely to result in mistakes than the git command line API. 

So our process should allow us to do PRs with the github GUI, as we have been doing until now.

**Comment by gold2718** [comment]
@edhartnett I would like to see if I understand your proposed workflow.

- 'develop' is always the next (in development) version of 'master'
- developers pull their branches from 'develop'
- 'develop' is tested regularly with Jenkins
- developers submit pull requests to 'develop'
----------------------------------------
- When a decision is made to update 'master', 'develop' is temporarily frozen
- When full test dashboard is green, 'develop' is merged to 'master'
- After merge to 'master', 'develop' handling of pending pull requests is resumed

Questions:
1. Does this look correct? If not, please submit a revised set.
2. When is history rewriting okay? For instance, one way to make sure pull requests are clean is to rebase my development branch against 'develop' and test that before submitting a pull request. Is that encouraged / allowed / forbidden in your proposed workflow?
3. If history rewriting is okay, is squashing of trivial and stupid commits encouraged / allowed / forbidden?

**Comment by edhartnett** [comment]
Yes, that is correct. But develop does not really have to be "frozen" for very long - only as long as it takes to test and merge to master. That could even be automated, to happen whenever develop changes and then passes all tests. But on the projects that used this, we just did it by hand.

It is encouraged to rebase and test before submitting a PR - you must do this for any conflicts before submitting code for review. Squashing of trivial comments is fine.

develop is always the next branch, unless something goes wrong with it, and then it is fixed before the next merge to master is done.

**Comment by gold2718** [comment]
Well, that meets my needs (stable 'master' + clear development workflow).
Are there stakeholders whose needs are not met with this workflow?
Is there a danger presented by this workflow which is likely given the size and scope of the project?

**Comment by edhartnett** [comment]
Let's use @rljacob's workflow for the rest of this week. I didn't really get it before, let me try again. I still may wish to change back to a simpler process next week, but at least let me give this one a proper try, and understand it.

**Comment by edhartnett** [comment]
I would like to suggest that we fall back to a simpler process for the next few months.

Then, if there is still interest, let's try the more complex process.

Both processes involve a develop and a master branch, so the simpler process will allow us to build up tests, jenkins builds, etc., and provide a very stable master.

**Comment by edhartnett** [comment]
Not sure why this workflow is popular with you guys.

Right now I cannot proceed because if I take my branch from master, and start making changes, there are no tickets left that I can work on without overlapping with changes I have already submitted in PR today.

So what is the answer to this issue?

**Comment by jedwards4b** [comment]
Take the branch from develop in this situation?   Let's discuss at the meeting this afternoon.

**Comment by edhartnett** [comment]
Once again held up by git issues.

I am trying to revert the last three commits to develop, but it does not work.

[ed@mikado ParallelIO]$ git checkout develop
Branch develop set up to track remote branch develop from origin.
Switched to a new branch 'develop'
[ed@mikado ParallelIO]$ git revert --no-commit HEAD~3..HEAD
error: Commit 960fe250de94b374f924e71ef860c9e539054345 is a merge but no -m option was given.
fatal: revert failed

**Comment by jedwards4b** [comment]
Since I think I made those commits I wonder why you want to revert them?   I checked the
difference between develop and master and was satisfied that I had it right.

**Comment by dqwu** [comment]
If you really want to revert, the following link might help:
http://stackoverflow.com/questions/5970889/why-does-git-revert-complain-about-a-missing-m-option

**Comment by edhartnett** [comment]
No @jedwards4b these are commits I made after you.

OK, I think I've straightened it out now.
